pkg/serviceaccount/serviceaccount_test.go

```diff
@@ -1 +1,328 @@
+			},
+		}
+	}
+
+	testCases := []struct {
+		saExistsAlready bool
+	}{
+		{
+			saExistsAlready: false,
+		},
+		{
+			saExistsAlready: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var runtimeObjects []runtime.Object
+
+		testSA := generateServiceAccount("test-sa", "test-ns")
+
+		if testCase.saExistsAlready {
+			runtimeObjects = append(runtimeObjects, testSA)
+		}
+
+		testBuilder := buildTestBuilderWithFakeObjects(runtimeObjects, testSA.Name, testSA.Namespace)
+		result, err := testBuilder.Create()
+		assert.Nil(t, err)
+		assert.Equal(t, testBuilder.Object.Name, result.Object.Name)
+	}
+}
+
+func TestServiceAccountDelete(t *testing.T) {
+	generateServiceAccount := func(name, namespace string) *corev1.ServiceAccount {
+		return &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: namespace,
+			},
+		}
+	}
+
+	testCases := []struct {
+		saExistsAlready bool
+	}{
+		{
+			saExistsAlready: false,
+		},
+		{
+			saExistsAlready: true,
+		},
+	}
+
+	for _, testCase := range testCases {
+		var runtimeObjects []runtime.Object
+
+		testSA := generateServiceAccount("test-sa", "test-ns")
+
+		if testCase.saExistsAlready {
+			runtimeObjects = append(runtimeObjects, testSA)
+		}
+
+		testBuilder := buildTestBuilderWithFakeObjects(runtimeObjects, testSA.Name, testSA.Namespace)
+		err := testBuilder.Delete()
+		assert.Nil(t, err)
+	}
+}
+
+func TestServiceAccountValidate(t *testing.T) {
+	testCases := []struct {
+		builderNil    bool
+		definitionNil bool
+		apiClientNil  bool
+		expectedError string
+	}{
+		{
+			builderNil:    true,
+			definitionNil: false,
+			apiClientNil:  false,
+			expectedError: "error: received nil ServiceAccount builder",
+		},
+		{
+			builderNil:    false,
+			definitionNil: true,
+			apiClientNil:  false,
+			expectedError: "can not redefine the undefined ServiceAccount",
+		},
+		{
+			builderNil:    false,
+			definitionNil: false,
+			apiClientNil:  true,
+			expectedError: "ServiceAccount builder cannot have nil apiClient",
+		},
+		{
+			builderNil:    false,
+			definitionNil: false,
+			apiClientNil:  false,
+			expectedError: "",
+		},
+	}
+
+	for _, testCase := range testCases {
+		testBuilder := buildValidTestBuilder()
+
+		if testCase.builderNil {
+			testBuilder = nil
+		}
+
+		if testCase.definitionNil {
+			testBuilder.Definition = nil
+		}
+
+		if testCase.apiClientNil {
+			testBuilder.apiClient = nil
+		}
+
+		result, err := testBuilder.validate()
+		if testCase.expectedError != "" {
+			assert.NotNil(t, err)
+			assert.Equal(t, testCase.expectedError, err.Error())
+			assert.False(t, result)
+		} else {
+			assert.Nil(t, err)
+			assert.True(t, result)
+		}
+	}
+}
+
+func TestServiceAccountWithOptions(t *testing.T) {
+	testBuilder := buildValidTestBuilder()
+
+	testBuilder.WithOptions(func(builder *Builder) (*Builder, error) {
+		return builder, nil
+	})
+
+	assert.Equal(t, "", testBuilder.errorMsg)
+}
+
+func buildValidTestBuilder() *Builder {
```

Comment:
Please move private function to the end of the file

Reply:
Done.